### DISCUSSION TITLE
Pause a query if there is no internet connection

### DISCUSSION THREAD

**Discussion by orjive5** [TanStack/query#7385] [Q&A]
How to prevent a query from resolving if the internet connection is lost while it's pending? Right now it returns an error, but I need it to be paused, invalidated or something. I'm new to react query, please help, thanks

**Comment by TkDodo** [comment]
what you should be seeing is that a query, which has 3 retries per default, goes into `paused` state if there is no network connection. Then, it will continue automatically once the connection is regained. At least those are the default settings, might be different if you've changed them.

**Reply by orjive5** [reply]
The query is initiated while there was a connection, and while it's pending, the connection is lost. It returns an error and doesn't retry. 
<img width="487" alt="Screenshot 2024-05-06 at 17 24 12" src="https://github.com/TanStack/query/assets/93201142/8f4d54cf-33f7-44a4-8cdf-5e7bf8fef756">

**Reply by TkDodo** [reply]
in this case, if `retry` is set (defaults to `3`), the query should go into paused state after the error has happened.

**Reply by antoantonydh** [reply]
@TkDodo I have network mode as always(using v4) as default config. but still query gets paused. is it because of the retry failure?

```
{
  dataUpdateCount: 0,
  dataUpdatedAt: 0,
  error: null,
  errorUpdateCount: 0,
  errorUpdatedAt: 0,
  fetchFailureCount: 1,
  fetchFailureReason: [Object],
  fetchMeta: null,
  fetchStatus: paused,
  isInvalidated: false,
  status: loading
}
```

and this how I configured the query


  ```
const {
    data,
    refetch,
    fetchStatus,
    isLoading,
    isFetching,
  } = useQuery<Todo[]>(
    [TEST_QUERY_KEY],
    async () => {
     try {
        const result = await getTodos();
        return result;
      } catch (error) {
        warning('Failed to load todos', error, {pollingIntervalInSeconds});
        throw error;
      }
    },
    {
      refetchInterval: pollingIntervalInSeconds * 1000,
      refetchIntervalInBackground: true,
      retry: 3,
    },
  );
```

**Reply by TkDodo** [reply]
`networkMode: 'always'` shouldn't pause:

https://github.com/TanStack/query/blob/18953be6f61553f4ed3dd888945dfa1683aa6e40/packages/query-core/src/retryer.ts#L97-L100

We _do_ pause if the window isn’t focused - given that you’ve set `refetchIntervalInBackground: true,`, this might be what’s happening here?

Otherwise, please show a minimal reproduction (stackblitz, codesandbox)

**Reply by antoantonydh** [reply]
> We do pause if the window isn’t focused - given that you’ve set refetchIntervalInBackground: true,, this might be what’s happening here?

Thanks a lot! Yes, this is indeed happening, and I was able to replicate it as well.

https://codesandbox.io/p/sandbox/suspicious-butterfly-gk6c6t

I initially assumed that setting the `refetchIntervalInBackground: true` configuration meant it would poll even when the tab is inactive(window isn’t focused). However, I’ve realized that refetching stops when a retry is in progress, and the retry doesn’t continue if the user is away from the tab.

We have users with unstable internet connections, where the internet frequently comes and goes. For these users, polling doesn’t occur because they get stuck in this retry + tab inactive scenario, even though their connection intermittently comes back. This has led to some issues on our end.

Is this behavior expected, or is it a bug? Shouldn’t retries continue regardless of the tab’s activity when `refetchIntervalInBackground` is set to true?

**Reply by antoantonydh** [reply]
I am planning to use the following way to override this behaviour:

```
import {focusManager, QueryClient} from '@tanstack/react-query';
focusManager.setFocused(true);
```

**Reply by TkDodo** [reply]
> Is this behavior expected, or is it a bug?

It has been here since before me, and there are tests for it, so I guess it’s expected.

If you have a `refetchInterval`, I think you might not want / need retries?

**Reply by antoantonydh** [reply]
Yeah, good point. But let's say we have a long polling interval of 5 minutes.

If, for some reason, the refetch fails on the first attempt due to a network issue or something similar, a retry could be helpful here.

Otherwise, we might not receive any data until the next refetch cycle, which could create a gap.

**Reply by TkDodo** [reply]
you can set a dynamic interval based on the `query` by implementing `refetchInterval: (query) => ...` and make it a smaller interval when the query is in error state.

**Reply by antoantonydh** [reply]
oh, nice. So we keep the retry off, then we reduce the `refetchInterval` dynamically based on the error scenario. Thanks, I will try this out.

**Comment by aryankholqi** [comment]
you can use this implementation:
  const { data, error, isLoading, refetch } = useQuery('myData', fetchData, {
    retry: false, // Disable automatic retry
    onError: (error) => {
      if (error.message === 'Network response was not ok') {
        // Handle network errors here
        // Display a message to the user indicating loss of internet connection
        console.error('Network error occurred');
      }
    },
  });
